File: solutions/dart/game-of-life/1/lib/game_of_life.dart
// In humble reverence to Shree DR.MDD
typedef Matrix = List<List<int>>;

class GameOfLife {
  Matrix _grid;

  GameOfLife(this._grid);

  Matrix matrix() => _grid;

  Matrix tick() {
    Matrix future = _blankGrid();

    for (var r = 0; r < _grid.length; r++) {
      for (var c = 0; c < _grid[r].length; c++) {
        future[r][c] = switch (_liveAround(r, c)) {
          3 => 1,
          2 => _grid[r][c],
          _ => 0,
        };
      }
    }

    _grid = future;
    return future;
  }

  Matrix _blankGrid() =>
      List.generate(_grid.length, (i) => List.filled(_grid[i].length, 0));

  int _liveAround(int r, int c) {
    var sum = 0;

    for (var offset in [
      [-1, -1], [-1, 0], [-1, 1],
      [0, -1],          [0, 1],
      [1, -1], [1, 0], [1, 1]
    ]) {
      var nr = r + offset[0];
      var nc = c + offset[1];
      if (nr >= 0 && nr < _grid.length &&
          nc >= 0 && nc < _grid[r].length) {
        sum += _grid[nr][nc];
      }
    }

    return sum;
  }
}

File: solutions/typescript/game-of-life/1/game-of-life.ts
// For Shree DR.MDD

export class GameOfLife {

  private _grid: number[][]
  
  constructor(matrix: number[][]) {
    this._grid = matrix
  }

  private countAliveNeighbours(x: number, y: number) {
    let aliveCount = 0
    
    for (let dy = -1; dy <= 1; dy++) {
      for (let dx = -1; dx <= 1; dx++) {
        if (this._grid[y + dy] && this._grid[y + dy][x + dx] === 1 && !(dx === 0 && dy === 0))
          aliveCount++
      }
    }

    return aliveCount
  }
  
  public tick() {
    const nextState = JSON.parse(JSON.stringify(this._grid))

    for (let row = 0; row < this._grid.length || 0; row++) {
      for (let col = 0; col < this._grid[0].length; col++) {
        let cellNext = 0
        const neighboursAlive = this.countAliveNeighbours(col, row)
        
        if (this._grid[row][col] === 1 && (neighboursAlive === 2 || neighboursAlive === 3))
          cellNext = 1
        if (this._grid[row][col] === 0 && neighboursAlive === 3)
          cellNext = 1
        nextState[row][col] = cellNext
      }
    }

    this._grid = nextState
  }

  public state() {
    return this._grid
  }
}
